segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerEventProcessorImpl.java

```diff
@@ -0,0 +1,500 @@
+                }
+            });
+            eventProcessorMap.clear();
+            log.debug("{}: Closing EventProcessors complete.", this.traceObjectId);
+        }
+    }
+
+    //endregion
+
+    //region ContainerEventProcessor Implementation
+
+    @Override
+    public CompletableFuture<EventProcessor> forConsumer(@NonNull String name,
+                                                         @NonNull Function<List<BufferView>, CompletableFuture<Void>> handler,
+                                                         @NonNull EventProcessorConfig config) {
+        Exceptions.checkNotClosed(this.closed.get(), this);
+        Preconditions.checkArgument(!name.isEmpty(), "EventProcessor name cannot be empty.");
+
+        synchronized (eventProcessorMap) {
+            // If the EventProcessor is already loaded, just return it.
+            if (eventProcessorMap.containsKey(name)) {
+                return CompletableFuture.completedFuture(eventProcessorMap.get(name));
+            }
+
+            // Instantiate the EventProcessor and put it into the map. If the EventProcessor is closed, auto-unregister.
+            Runnable onClose = () -> eventProcessorMap.remove(name);
+            return segmentSupplier.apply(name)
+                    .thenApply(segment -> {
+                        Exceptions.checkNotClosed(this.closed.get(), this);
+                        EventProcessorImpl processor = new EventProcessorImpl(name, segment, handler, config, onClose, executor);
+                        processor.startAsync().awaitRunning();
+                        eventProcessorMap.put(name, processor);
+                        return processor;
+                    });
+        }
+    }
+
+    @Override
+    public CompletableFuture<EventProcessor> forDurableQueue(@NonNull String name) {
+        Exceptions.checkNotClosed(this.closed.get(), this);
+        Preconditions.checkArgument(!name.isEmpty(), "EventProcessor name cannot be empty.");
+
+        // Do not limit the amount of outstanding bytes when there is no consumer configured.
+        EventProcessorConfig config = new ContainerEventProcessor.EventProcessorConfig(0, Long.MAX_VALUE);
+        synchronized (eventProcessorMap) {
+            // If the EventProcessor is already loaded, just return it.
+            if (eventProcessorMap.containsKey(name)) {
+                return CompletableFuture.completedFuture(eventProcessorMap.get(name));
+            }
+
+            // Instantiate the EventProcessor and put it into the map. If the EventProcessor is closed, auto-unregister.
+            Runnable onClose = () -> eventProcessorMap.remove(name);
+            Function<List<BufferView>, CompletableFuture<Void>> noHandler = l -> CompletableFuture.completedFuture(null);
+            return segmentSupplier.apply(name)
+                    .thenApply(segment -> {
+                        Exceptions.checkNotClosed(this.closed.get(), this);
```

Comment:
Actually this code will not execute under the lock!

This is a callback that is called when `segmentSupplier.apply` returns; by the time it does, it will likely be on a different thread and the original lock will have been released.

So this code does not do what you think it does  - taking care of concurrent calls. 

See how we've done it in other places (look at `ContainerAttributeIndex`) . We record a future and latch on to it. Any subsequent callers will get the result of the original call.

Reply:
That's true, good catch. As suggested, I have used a latch that will release when the `EventProcessor` is instantiated and stored in the map. When the latch is released, the `forConsumer` and `forDurableQueue` will exist the synchronized block. I think that this should prevent any concurrent accesses to that method. Also, the main thread in the synchronized block will block only for some time (`containerOperationTimeout` millis). If the instantiation of the `EventProcessor` takes too long, the future returned to the client will complete exceptionally (added a new test to verify this).